lib/rex/proto/ldap/auth.rb

```diff
@@ -0,0 +1,329 @@
+              pub_info = user_login.name.split('\\')
+              if pub_info.length <= 2
+                auth_info[:domain], auth_info[:user] = pub_info
+              else
+                auth_info[:result_code] = Net::LDAP::ResultCodeInvalidCredentials
+                auth_info[:error_msg] = "Invalid LDAP Login Attempt => DN:#{user_login.name}"
+              end
+            else
+              auth_info[:user] = user_login.name
+              auth_info[:domain] = nil
+              auth_info[:result_code] = Net::LDAP::ResultCodeInvalidCredentials
+            end
+            auth_info[:private] = user_login.authentication
+            auth_info[:private_type] = :password
+            auth_info[:result_code] = Net::LDAP::ResultCodeAuthMethodNotSupported if auth_info[:result_code].nil?
+            auth_info[:auth_type] = 'Simple'
+            auth_info
+          end
+        end
+
+        #
+        # Handle SASL authentication requests
+        #
+        # @param user_login [OpenStruct] User login information
+        # @param auth_info [Hash] Processed authentication information
+        #
+        # @return auth_info [Hash] Processed authentication information
+        def handle_sasl_request(user_login, auth_info = {})
+          if user_login.authentication[1] =~ /NTLMSSP/
+            message = Net::NTLM::Message.parse(user_login.authentication[1])
+
+            if message.is_a?(::Net::NTLM::Message::Type1)
+              auth_info[:server_creds] = generate_type2_response(message)
+              auth_info[:result_code] = Net::LDAP::ResultCodeSaslBindInProgress
+            elsif message.is_a?(::Net::NTLM::Message::Type3)
+              auth_info = handle_type3_message(message, auth_info)
+              auth_info[:result_code] = Net::LDAP::ResultCodeAuthMethodNotSupported
+            end
+          end
+          auth_info[:auth_type] = 'SASL'
+          auth_info
+        end
+
+        private
+
+        #
+        # Determine if the supplied request is formatted for SASL auth
+        #
+        # @param user_login [OpenStruct] User login information
+        #
+        # @return [bool] True if the request can be processed for SASL auth
+        def sasl?(user_login)
+          if user_login.authentication.is_a?(Array) && SUPPORTS_SASL.include?(user_login.authentication[0])
+            return true
+          end
+
+          false
+        end
+
+        #
+        # Generate NTLM Type2 response from NTLM Type1 message
+        #
+        # @param message [Net::NTLM::Message::Type1] NTLM Type1 message
+        #
+        # @return server_hash [String] NTLM Type2 response that is sent as server credentials
+        def generate_type2_response(message)
+          dom = message.domain
+          ws = message.workstation
+          @domain = dom if dom
+          @server = ws if ws
+          server_hash = MESSAGE.process_type1_message(message.encode64, @challenge, @domain, @server, @dnsname, @dnsdomain)
```

Comment:
This value is calculated but never used. Is this intentional?

Reply:
See line 176 this value is base64 encoded as the return value for this method.

Reply:
I will expand the rspec to try and clarify this with a test:
```
expect(Net::NTLM::Message.parse(result[:server_creds])).to(be_a(Net::NTLM::Message::Type2))
```

Reply:
On further inspection looks like a change brought in extra code here that should have been removed, addressed in 4cb18483d6